toolchain/semantics/semantics_context.cpp

```diff
@@ -76,43 +78,93 @@ auto SemanticsContext::AddNodeAndPush(ParseTree::Node parse_node,
+      // The name is invalid and a diagnostic has already been emitted.
+      return;
+
+    case DeclarationNameContext::State::New:
+      CARBON_FATAL() << "Name is missing, not expected to call AddNameToLookup "
+                        "(but that may change based on error handling).";
+
+    case DeclarationNameContext::State::Resolved:
+    case DeclarationNameContext::State::ResolvedNonScope: {
+      auto prev_def = semantics_ir_->GetNode(name_context.resolved_node_id);
+      emitter_->Build(name_context.parse_node, NameDeclarationDuplicate)
+          .Note(prev_def.parse_node(), NameDeclarationPrevious)
+          .Emit();
+      return;
+    }
 
-  auto name_id = semantics_ir_->GetStringID(name);
-  if (!name_id) {
-    emitter_->Emit(parse_node, NameNotFound, name);
-    return SemanticsNodeId::BuiltinInvalidType;
+    case DeclarationNameContext::State::Unresolved:
+      if (name_context.target_scope_id == SemanticsNameScopeId::Invalid) {
+        AddNameToLookup(name_context.parse_node,
+                        name_context.unresolved_name_id, target_id);
+      } else {
+        bool success = semantics_ir_->AddNameScopeEntry(
+            name_context.target_scope_id, name_context.unresolved_name_id,
+            target_id);
+        CARBON_CHECK(success)
+            << "Duplicate names should have been resolved previously: "
+            << name_context.unresolved_name_id << " in "
+            << name_context.target_scope_id;
+      }
+      return;
   }
+}
 
-  auto it = name_lookup_.find(*name_id);
-  if (it == name_lookup_.end()) {
-    emitter_->Emit(parse_node, NameNotFound, name);
-    return SemanticsNodeId::BuiltinInvalidType;
-  }
-  CARBON_CHECK(!it->second.empty()) << "Should have been erased: " << name;
+auto SemanticsContext::LookupName(ParseTree::Node parse_node,
+                                  SemanticsStringId name_id,
+                                  SemanticsNameScopeId scope_id,
+                                  bool print_diagnostics) -> SemanticsNodeId {
+  if (scope_id == SemanticsNameScopeId::Invalid) {
+    auto it = name_lookup_.find(name_id);
+    if (it == name_lookup_.end()) {
+      if (print_diagnostics) {
+        emitter_->Emit(parse_node, NameNotFound,
+                       semantics_ir_->GetString(name_id));
+      }
+      return SemanticsNodeId::BuiltinInvalidType;
+    }
+    CARBON_CHECK(!it->second.empty())
+        << "Should have been erased: " << semantics_ir_->GetString(name_id);
 
-  // TODO: Check for ambiguous lookups.
-  return it->second.back();
+    // TODO: Check for ambiguous lookups.
+    return it->second.back();
+  } else {
+    auto scope = semantics_ir_->GetNameScope(scope_id);
+    auto it = scope.find(name_id);
+    if (it == scope.end()) {
+      if (print_diagnostics) {
+        emitter_->Emit(parse_node, NameNotFound,
+                       semantics_ir_->GetString(name_id));
+      }
+      return SemanticsNodeId::BuiltinInvalidType;
```

Comment:
The InvalidType is being used as a type in semantics (it's specifically valid as a type_id). Given the directionality (types can be values, but values have a hard time becoming types), I prefer the type label over value.

We could also break out of the type/value mindset and go for something like "SemanticError" which may better capture the concept -- I've been mulling a shift in that direction mainly to make the distinction clearer, that this isn't _really_ a type.

Reply:
Perhaps a different viewpoint would be that this value is permitted basically anywhere as a placeholder for an error, which is why we allow it as a `type_id` in particular. We'd also presumably allow this value as a base class (despite it not being a class), or as an array bound (despite it not being an integer), so it seems consistent to allow it as a type (despite it not being a value of type `type`).

`SemanticError` seems fine to me. I guess that'd be broader than `InvalidValue`, in that it'd potentially be usable even in non-value contexts where some other kind of node is typically expected. That probably adds some complexity to consumers of those other kinds of node, though, if they otherwise only had one semantic node kind that they needed to handle.

Reply:
To be sure, still as a builtin. The complexity of needing to handle errors is, I think, there regardless of the shape it takes. However, not using a node kind reduces a level of investigation needed to detect an error.